go/arrow/avro/reader_types.go

```diff
@@ -0,0 +1,892 @@
+	var child *dataLoader = &dataLoader{
+		depth: d.depth + 1,
+	}
+	d.children = append(d.children, child)
+	return child
+}
+
+func (d *dataLoader) newListChild(list *fieldPos) *dataLoader {
+	var child *dataLoader = &dataLoader{
+		list:  list,
+		item:  list.childrens[0],
+		depth: d.depth + 1,
+	}
+	d.children = append(d.children, child)
+	return child
+}
+
+func (d *dataLoader) newMapChild(mapField *fieldPos) *dataLoader {
+	var child *dataLoader = &dataLoader{
+		mapField: mapField,
+		depth:    d.depth + 1,
+	}
+	d.children = append(d.children, child)
+	return child
+}
+
+type fieldPos struct {
+	parent       *fieldPos
+	fieldName    string
+	builder      array.Builder
+	path         []string
+	isList       bool
+	isItem       bool
+	isStruct     bool
+	isMap        bool
+	typeName     string
+	appendFunc   func(val interface{}) error
+	metadatas    arrow.Metadata
+	childrens    []*fieldPos
+	index, depth int32
+}
+
+func newFieldPos() *fieldPos { return &fieldPos{index: -1} }
+
+func (f *fieldPos) name() string { return f.fieldName }
+
+func (f *fieldPos) child(index int) (*fieldPos, error) {
+	if index < len(f.children()) {
+		return f.childrens[index], nil
+	}
+	return nil, fmt.Errorf("%v child index %d not found", f.namePath(), index)
+}
+
+func (f *fieldPos) children() []*fieldPos { return f.childrens }
+
+func (f *fieldPos) metadata() arrow.Metadata { return f.metadatas }
+
+func (f *fieldPos) newChild(childName string, childBuilder array.Builder, meta arrow.Metadata) *fieldPos {
+	var child fieldPos = fieldPos{
+		parent:    f,
+		fieldName: childName,
+		builder:   childBuilder,
+		metadatas: meta,
+		index:     int32(len(f.childrens)),
+		depth:     f.depth + 1,
+	}
+	if f.isList {
+		child.isItem = true
+	}
+	child.path = child.namePath()
+	f.childrens = append(f.childrens, &child)
+	return &child
+}
+
+// NamePath returns a slice of keys making up the path to the field
+func (f *fieldPos) namePath() []string {
+	if len(f.path) == 0 {
+		var path []string
+		var listPath []string
+		cur := f
+		for i := f.depth - 1; i >= 0; i-- {
+			if cur.typeName == "" {
+				path = append([]string{cur.fieldName}, path...)
+			} else {
+				path = append([]string{cur.fieldName, cur.typeName}, path...)
+			}
+			if !cur.parent.isMap {
+				cur = cur.parent
+			}
+		}
+		if f.parent.parent != nil && f.parent.parent.isList {
+			for i := len(path) - 1; i >= 0; i-- {
+				if path[i] != "item" {
+					listPath = append([]string{path[i]}, listPath...)
+				} else {
+					return listPath
+				}
+			}
+		}
+		if f.parent != nil && f.parent.fieldName == "value" {
+			for i := len(path) - 1; i >= 0; i-- {
+				if path[i] != "value" {
+					listPath = append([]string{path[i]}, listPath...)
+				} else {
+					return listPath
+				}
+			}
+		}
+		return path
+	}
+	return f.path
+}
+
+// GetValue retrieves the value from the map[string]any
+// by following the field's key path
+func (f *fieldPos) getValue(m any) any {
+	var value any = m
```

Comment:
why bother with this instead of either just naming the parameter `value` or using `switch value := m.(type)` or just using `m` below? This seems to be unneccessary

Reply:
used m below and removed variable assignment